core/src/session.rs

```diff
@@ -187,19 +187,30 @@ impl Session {
         credentials: Credentials,
         store_credentials: bool,
     ) -> Result<(), Error> {
-        let (reusable_credentials, transport) = loop {
+        const MAX_RETRIES: u8 = 2;
+        let (reusable_credentials, transport) = 'ap: loop {
             let ap = self.apresolver().resolve("accesspoint").await?;
             info!("Connecting to AP \"{}:{}\"", ap.0, ap.1);
-            match self.connect_inner(ap, credentials.clone()).await {
-                Ok(ct) => break ct,
-                Err(e) => {
-                    if let Some(AuthenticationError::LoginFailed(ErrorCode::TryAnotherAP)) =
-                        e.error.downcast_ref::<AuthenticationError>()
-                    {
-                        warn!("Instructed to try another access point...");
-                        continue;
-                    } else {
-                        return Err(e);
+            let mut num_retries = 0;
+            loop {
+                match self.connect_inner(&ap, credentials.clone()).await {
+                    Ok(ct) => break 'ap ct,
+                    Err(e) => {
+                        if let Some(AuthenticationError::LoginFailed(ErrorCode::TryAnotherAP)) =
+                            e.error.downcast_ref::<AuthenticationError>()
+                        {
+                            warn!("Instructed to try another access point...");
+                            continue 'ap;
+                        } else {
+                            debug!("Connection failed: {e}");
+                            if num_retries == MAX_RETRIES {
+                                warn!("Retries exceeded, try another access point...");
+                                continue 'ap;
+                            } else {
+                                num_retries += 1;
+                                trace!("Retry access point...");
```

Comment:
I guess that's the message I'm looking for to see this fallback in action?

```
[2024-09-20T04:30:53Z DEBUG librespot::component] new ApResolver
[2024-09-20T04:30:53Z DEBUG librespot_core::http_client] Requesting https://apresolve.spotify.com/?type=accesspoint&type=dealer&type=spclient
[2024-09-20T04:30:54Z INFO  librespot_core::session] Connecting to AP "ap-gew4.spotify.com:4070"
[2024-09-20T04:30:54Z DEBUG librespot_core::session] Connection failed: Service unavailable { Connection refused (os error 61) }
[2024-09-20T04:30:54Z TRACE librespot_core::session] Retry access point...
[2024-09-20T04:30:54Z DEBUG librespot_core::connection] Authenticating with AP using AUTHENTICATION_STORED_SPOTIFY_CREDENTIALS
[2024-09-20T04:30:54Z INFO  librespot_core::session] Authenticated as 'me' !
```

Seems to have caught a first failure successfully! 👍🏻